tsa-core/src/main/kotlin/org/usvm/machine/interpreter/TvmDictOperationInterpreter.kt

```diff
@@ -1637,18 +1929,154 @@ class TvmDictOperationInterpreter(
         )
     }
 
+    private fun handleDeleteInstOnInputDict(
+        scope: TvmStepScopeManager,
+        dictCellRef: UHeapRef,
+        keySort: KBvSort,
+        keyKind: DictKeyKind,
+        key: UExpr<UBvSort>,
+        dictId: DictId,
+        valueType: DictValueType,
+        getOldValue: Boolean,
+        inst: TvmDictDeleteInst,
+    ) {
+        val resultDictRef = scope.calcOnState { memory.allocConcrete(TvmDictCellType) }
+        val dictOriginalConcreteRef =
+            dictCellRef as? UConcreteHeapRef
+                ?: TODO("unsupported")
+        val initInputDict =
+            scope.calcOnState {
+                readInputDictionary(dictOriginalConcreteRef, keySort, keyKind)
+            }
+
+        val queryKeyExists =
+            doInputDictHasKey(scope, initInputDict, KeyType(key, keyKind))?.exists
+                ?: return
+        val appliedModification = Modification.Remove(KeyType(key, keyKind))
+        val newInputDict = initInputDict.withModification(appliedModification)
+        scope.calcOnStateCtx {
+            copyInputDictMemoryRepresentation(dictOriginalConcreteRef, resultDictRef, dictId, keySort)
+            inputDictionaryStorage = inputDictionaryStorage.createDictEntry(resultDictRef, newInputDict)
+        }
+
+        // assert not empty
+        val someKeyExistsAfterDeletion =
+            doInputDictHasKey(
+                scope,
+                newInputDict,
+                scope.calcOnState { makeFreshKeyConstant(keySort, keyKind) },
+            )?.exists
+                ?: return
+        assertInputDictIsEmpty(
+            scope,
+            newInputDict,
+            ctx.mkNot(someKeyExistsAfterDeletion),
+        ) ?: return
+
+        // TODO: once the ite operations on the dictionary are fully supported,
+        // replace this with ITE reference
+        val actions =
+            listOf(
+                TvmStepScopeManager.ActionOnCondition(
+                    action = {},
+                    caseIsExceptional = false,
+                    condition = someKeyExistsAfterDeletion,
+                    paramForDoForAllBlock = resultDictRef,
+                ),
+                TvmStepScopeManager.ActionOnCondition(
+                    action = {},
+                    caseIsExceptional = false,
+                    condition = ctx.mkNot(someKeyExistsAfterDeletion),
+                    paramForDoForAllBlock = ctx.nullValue,
+                ),
+            )
```

Comment:
This block is misleading. Don't you fork on each variant? If so, why do you assert both conditions? If you assert contradicting constraints, then your state must die.

Do you have tests for this case?